src/rooms/room_details/v1.rs

```diff
@@ -0,0 +1,142 @@
+//! [GET /_synapse/admin/v1/rooms/:room_id](https://github.com/element-hq/synapse/blob/master/docs/admin_api/rooms.md#room-details-api)
+use ruma::{
+    api::{metadata, request, response, Metadata},
+    events::room::{guest_access::GuestAccess, history_visibility::HistoryVisibility},
+    room::RoomType,
+    space::SpaceRoomJoinRule,
+    OwnedRoomAliasId, OwnedRoomId, OwnedUserId, UInt,
+};
+
+const METADATA: Metadata = metadata! {
+    method: GET,
+    rate_limited: false,
+    authentication: AccessToken,
+    history: {
+        unstable => "/_synapse/admin/v1/rooms/:room_id",
+    }
+};
+
+#[request]
+pub struct Request {
+    /// Alias or ID of the room to join.
+    #[ruma_api(path)]
+    pub room_id: OwnedRoomId,
+}
+
+#[response]
+pub struct Response {
+    /// Room ID
+    pub room_id: OwnedRoomId,
+
+    /// Room name
+    pub name: Option<String>,
+
+    /// Room topic
+    pub topic: Option<String>,
+
+    /// Room avatar
+    pub avatar: Option<String>,
+
+    /// Room alias ID
+    pub canonical_alias: Option<OwnedRoomAliasId>,
+
+    /// Amount of joined members.
+    pub joined_members: UInt,
+
+    /// Amount of local members.
+    pub joined_local_members: UInt,
+
+    ///A mount of local devices.
+    pub joined_local_devices: UInt,
+
+    /// Room version
+    pub version: String,
+
+    /// User ID of the room creator.
+    #[serde(deserialize_with = "ruma::serde::empty_string_as_none")]
+    pub creator: Option<OwnedUserId>,
+
+    /// Room encryption.
+    pub encryption: Option<String>,
+
+    /// Whether the room is federatable
+    #[serde(deserialize_with = "crate::serde::bool_or_uint")]
+    pub federatable: bool,
+
+    /// Whether the room is public.
+    #[serde(deserialize_with = "crate::serde::bool_or_uint")]
+    pub public: bool,
+
+    /// Join rules of the room.
+    pub join_rules: Option<SpaceRoomJoinRule>,
+
+    /// Guest access of the room
+    pub guest_access: Option<GuestAccess>,
+
+    /// History visibility of the room
+    pub history_visibility: Option<HistoryVisibility>,
+
+    /// State events of the room.
+    pub state_events: UInt,
+
+    /// Room type of the room.
+    pub room_type: Option<RoomType>,
+
+    /// Whether all local users have forgotten the room.
+    #[serde(deserialize_with = "crate::serde::bool_or_uint")]
+    pub forgotten: bool,
+}
+
+impl Request {
+    /// Creates an empty `Request`.
+    pub fn new(room_id: OwnedRoomId) -> Self {
+        Self { room_id }
+    }
+}
+
+impl Response {
+    /// Creates an empty `Response`.
+    pub fn new(
```

Comment:
clippy is right, this is not exactly a nice function signature. For this sort of situation we usually create an "Init" struct that can be constructed with a struct literal in Ruma, like [here](https://github.com/ruma/ruma/blob/5c9ef9f425d8f8b330e46d215c683bf799a28d50/crates/ruma-client-api/src/space.rs#L85).

Reply:
Why do we even bother creating the constructor? It works without as well?

Reply:
The `request` and `response` macros add the `non_exhaustive` attribute to the structs, unless the `ruma_unstable_exhaustive_types` `cfg` settings is set.

This attribute has the benefit of allowing to add new fields without it being a breaking change. However it also makes it impossible to construct the struct without a constructor.

Reply:
It's true that in theory only a server will need to construct this. But for the sake of completeness, we should probably add a constructor. I believe that a constructor that takes the `room_id` and uses the default value for all other fields could be enough.

Reply:
I can live with that. I added one, though I'm not sure if there is a cleaner way doing that.

Reply:
No that looks fine. Maybe if would be cleaner for the numbers to be constructed with `UInt::default()`, but I don't really mind.